### ISSUE TITLE
Error from cached load() is passed to the createAsync result as value

### ISSUE THREAD

**Issue by madaxen86** [solidjs/solid-router#385]
### Describe the bug

May be related to [#374](https://github.com/solidjs/solid-router/issues/374)
When the function that is passed to the load() throws an error or rejected promise, the error will be passed to the result of the createAsync primitive as successful value if the page is refreshed and then clicked on a link to the current page (please check the video).

### Your Example Website or App

https://codesandbox.io/p/devbox/black-forest-rvp7xj?workspaceId=55b47a62-ca2d-456f-9cef-0a44c292a811&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt62eq4r00053b6kwmsynsmf%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt62eq4r00023b6kr57gddyu%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt62eq4r00033b6kltnrrsbo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt62eq4r00043b6kkthi0q6d%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt62eq4r00023b6kr57gddyu%2522%253A%257B%2522id%2522%253A%2522clt62eq4r00023b6kr57gddyu%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clt62eq4r00043b6kkthi0q6d%2522%253A%257B%2522id%2522%253A%2522clt62eq4r00043b6kkthi0q6d%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clt63ch5v012g3b6k2nzu1onr%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252Ftest%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt63ch5v012g3b6k2nzu1onr%2522%257D%252C%2522clt62eq4r00033b6kltnrrsbo%2522%253A%257B%2522id%2522%253A%2522clt62eq4r00033b6kltnrrsbo%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt635se200ee3b6kwkx7pgtz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clt635sgp000ydjgrbgyw9chy%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clt635se200ee3b6kwkx7pgtz%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

### Steps to Reproduce the Bug or Issue

1. Open code sandbox
2.  Navigate between test and home page -->all good. Errorboundary is rendered
3. Go to test page
4. refresh the window
5. click on test navlink --> page is rendered with the data primitive containing the error string !?

### Expected behavior

The error should always be picked up by the ErrorBoundary and never passed to the primitive returned by the createAsync function.

### Screenshots or Videos

https://github.com/solidjs/solid-router/assets/59310516/d855df9c-35a2-4d46-b387-7c64a1134f56

### Platform

@solidjs/router:   0.12.4
@solidjs/start:  0.6.0

**Comment by ryansolid** [comment]
Yeah this one is odd.. codesandbox is not being particularly helpful to me because they have a limit now on drafts unless I pay so I can't play with this much other than confirm the reproduction and that it isn't the same as the other issue. Beyond that this one is very specific to loading on the page... so there probably is an issue related to cache + hydration. I will try to look into this more in a bit.

**Comment by madaxen86** [comment]
I added a stackblitz - here the error is on the home page because stackblitz reloads o on the home route when reloading the url.
Added a Github repo so you can dig into it. Thanks a lot and thanks for this amazing framework. 
It is such a pleasure to work with SolidJs.

**Comment by apatrida** [comment]
in 1.0 solid-start this seems to still be around:

I see other types like redirects being returned to createAsync causing it to blow up on exceptions as well.  Just throw a redirect from a server function (not wrapped in cache) that is received in createAsync and it blows up on the result type not being a string or something it recognizes.

```
async function serverApiCallThrow(isIndirect?: boolean): Promise<string> {
    'use server';
    throw redirect(`/${isIndirect ? 'fromIndirect' : 'fromThrow'}/whateva?q=456`);
    return 'ok'
}

...
  const asyncThing = createAsync(()=>serverApiCallThrow())
  
...
  <Show when={asyncThing()}>
      <div>boo!</div>
  </Show>
```

results in exception:

> Error: Unknown error at castError (...server bundle...)

and client shows:

<img width="849" alt="Screenshot 2024-05-26 at 8 23 10 AM" src="https://github.com/solidjs/solid-router/assets/182340/781eebfb-5160-4a72-ab23-2c59e9b5cf37">

**Comment by ryansolid** [comment]
Closing this one out as part of an audit of older cache issues, since it's no longer reproducible on the current codebase as far as we can determine.

The report predates several rewrites of this exact area: 884b8be (fix #374/#399) reworked how errors from `preload` interact with the cache so preload-time consumption doesn't swallow or convert rejections, 97184e4 relaxed cache resuming during hydration, and the `cache`->`query` handling of rejected entries has been rewritten since 0.12.4. I've also added a regression test (f340b6f) covering the reachable variant of this scenario - a `query` that rejects, rendered via `createAsync` + `ErrorBoundary`, navigated away from and back within the cache window - and the cached rejection stays an error surfaced through the boundary; it never leaks into the accessor as a value.

If you can still reproduce the SSR/hydration variant (refresh on the erroring page, then navigate) against `@solidjs/router` 0.16.x, please open a fresh issue with the repro and we'll dig in - a rejected serialized promise turning into a resolved value on the client would be a real bug, but nothing in the current pipeline appears to do that.
